packages/core/src/domain/session/sessionCookieStore.ts

```diff
@@ -1,27 +1,76 @@
 import { CookieOptions, getCookie, setCookie } from '../../browser/cookie'
 import * as utils from '../../tools/utils'
+import { isExperimentalFeatureEnabled } from '../configuration/experimentalFeatures'
+import { monitor } from '../internalMonitoring/internalMonitoring'
 import { SessionState, SESSION_EXPIRATION_DELAY } from './sessionStore'
 
 const SESSION_ENTRY_REGEXP = /^([a-z]+)=([a-z0-9-]+)$/
 const SESSION_ENTRY_SEPARATOR = '&'
 
 export const SESSION_COOKIE_NAME = '_dd_s'
+const RETRY_DELAY = 1
 
-export function withCookieLockAccess({
-  process,
-  after,
-  options
-}: {
-  process: (cookieSession: SessionState) => SessionState | undefined
-  after?: (cookieSession: SessionState) => void,
+type Operations = {
   options: CookieOptions
-}) {
-  const initialSession = retrieveSession()
-  const processedSession = process(initialSession)
+  process: (cookieSession: SessionState) => SessionState | undefined
+  after?: (cookieSession: SessionState) => void
+}
+
+export function withCookieLockAccess(operations: Operations) {
+  let currentLock: string
+  let currentSession = retrieveSession()
+  if (isExperimentalFeatureEnabled('cookie-lock')) {
+    // if someone has lock, postpone
+    if (currentSession.lock) {
+      postpone(operations)
+      return
+    }
+    // acquire lock
+    currentLock = utils.generateUUID()
+    currentSession.lock = currentLock
+    setSession(currentSession, operations.options)
+    // if lock is not acquired, postpone
+    currentSession = retrieveSession()
+    if (currentSession.lock !== currentLock) {
+      postpone(operations)
+      return
+    }
+  }
+  let processedSession = operations.process(currentSession)
+  if (isExperimentalFeatureEnabled('cookie-lock')) {
+    // if lock corrupted after process, postpone
+    currentSession = retrieveSession()
+    if (currentSession.lock !== currentLock!) {
+      postpone(operations)
+      return
+    }
+  }
   if (processedSession) {
-    persistSession(processedSession, options)
+    persistSession(processedSession, operations.options)
+  }
+  if (isExperimentalFeatureEnabled('cookie-lock')) {
+    if (!processedSession || !utils.isEmptyObject(processedSession)) {
+      // if lock corrupted after persist, postpone
+      currentSession = retrieveSession()
+      if (currentSession.lock !== currentLock!) {
+        postpone(operations)
+        return
+      }
+      delete currentSession.lock
+      setSession(currentSession, operations.options)
+      processedSession = currentSession
+    }
   }
-  after?.(processedSession || initialSession)
+  operations.after?.(processedSession || currentSession)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏Add a comment to justify calling `after` even if `process` returns undefined. FMU this is to handle the case when the session has been modified in another tab?

Reply:
> FMU this is to handle the case when the session has been modified in another tab?

indeed, with `watchSession`

added a comment